File: PHASE_3/Jacobi_MPI/jacobi_2d_full.cpp
// =============================================================
// FULL 2D JACOBI — 2D Decomposition with Cartesian Topology
// =============================================================
// Complete version featuring:
//  - MPI_Cart_create (2D process grid)
//  - MPI_Dims_create (automatic decomposition)
//  - Halo exchange on all 4 sides
//  - Convergence check with MPI_Allreduce
//
// Each process manages a rectangular subdomain.
//
// Compilation:  mpicxx -O2 -Wall -o jacobi2d jacobi_2d_full.cpp
// Execution:    mpirun -np 4 ./jacobi2d
// =============================================================

#include <mpi.h>
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <algorithm>

int main(int argc, char* argv[]) {
    MPI_Init(&argc, &argv);

    int rank_world, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank_world);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // Global parameters
    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    const int N        = 128;   // global N x N grid
    const double EPS   = 1e-5;
    const int MAX_ITER = 50000;

    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // 2D Cartesian topology
    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    int dims[2]    = {0, 0};   // computed by MPI_Dims_create
    int periods[2] = {0, 0};   // non-periodic boundaries

    MPI_Dims_create(size, 2, dims);

    MPI_Comm comm_cart;
    MPI_Cart_create(MPI_COMM_WORLD, 2, dims, periods, 1, &comm_cart);

    int rank;
    MPI_Comm_rank(comm_cart, &rank);

    int coords[2];
    MPI_Cart_coords(comm_cart, rank, 2, coords);

    int my_row = coords[0];  // row position in the process grid
    int my_col = coords[1];  // column position

    // Number of rows/columns handled by each process
    // (for simplicity, assume N is divisible by dims)
    int local_rows = N / dims[0];
    int local_cols = N / dims[1];

    // First global cell handled by this process
    int global_row_start = my_row * local_rows;
    int global_col_start = my_col * local_cols;

    if (rank == 0) {
        std::cout << "Global grid: " << N << "x" << N << "\n";
        std::cout << "Processes: " << dims[0] << "x" << dims[1]
                  << " (" << size << " total)\n";
        std::cout << "Subdomain per process: "
                  << local_rows << "x" << local_cols << "\n\n";
    }

    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // Neighbors (may be MPI_PROC_NULL on boundaries)
    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    int north, south, west, east, dummy;

    MPI_Cart_shift(comm_cart, 0, -1, &dummy, &north);
    MPI_Cart_shift(comm_cart, 0, +1, &dummy, &south);
    MPI_Cart_shift(comm_cart, 1, -1, &dummy, &west);
    MPI_Cart_shift(comm_cart, 1, +1, &dummy, &east);

    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // Allocation: subdomain + ghost cells
    // (1 ghost cell on each side)
    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    int nr = local_rows + 2;  // rows including ghosts
    int nc = local_cols + 2;  // columns including ghosts

    std::vector<double> u(nr * nc, 0.0);
    std::vector<double> u_new(nr * nc, 0.0);

    auto idx = [&](int i, int j) { return i * nc + j; };

    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // Boundary conditions: bottom boundary = 1.0
    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    bool has_south_boundary = (my_row == dims[0] - 1);

    if (has_south_boundary) {
        for (int j = 1; j <= local_cols; j++)
            u[idx(local_rows, j)] = 1.0;
    }

    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // MPI datatype for columns (horizontal halo exchange)
    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // Column data is not contiguous in memory,
    // so we create a derived MPI datatype.
    MPI_Datatype col_type;
    MPI_Type_vector(local_rows, 1, nc, MPI_DOUBLE, &col_type);
    MPI_Type_commit(&col_type);

    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    // Jacobi loop
    // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    double t0 = MPI_Wtime();
    int iter = 0;
    double global_change = 1.0;

    const int TAG1 = 1, TAG2 = 2, TAG3 = 3, TAG4 = 4;

    while (global_change > EPS && iter < MAX_ITER) {

        // --- North-South Halo Exchange (rows) ---
        MPI_Sendrecv(&u[idx(1, 1)],          local_cols, MPI_DOUBLE, north, TAG1,
                     &u[idx(0, 1)],          local_cols, MPI_DOUBLE, north, TAG2,
                     comm_cart, MPI_STATUS_IGNORE);

        MPI_Sendrecv(&u[idx(local_rows, 1)], local_cols, MPI_DOUBLE, south, TAG2,
                     &u[idx(nr - 1, 1)],     local_cols, MPI_DOUBLE, south, TAG1,
                     comm_cart, MPI_STATUS_IGNORE);

        // --- East-West Halo Exchange (columns using derived datatype) ---
        MPI_Sendrecv(&u[idx(1, 1)],          1, col_type, west, TAG3,
                     &u[idx(1, nc - 1)],     1, col_type, east, TAG3,
                     comm_cart, MPI_STATUS_IGNORE);

        MPI_Sendrecv(&u[idx(1, local_cols)], 1, col_type, east, TAG4,
                     &u[idx(1, 0)],          1, col_type, west, TAG4,
                     comm_cart, MPI_STATUS_IGNORE);

        // --- Jacobi Update ---
        double local_change = 0.0;

        for (int i = 1; i <= local_rows; i++) {
            int ig = global_row_start + (i - 1);

            for (int j = 1; j <= local_cols; j++) {
                int jg = global_col_start + (j - 1);

                // Preserve boundary conditions
                if (ig == 0 || ig == N - 1 || jg == 0 || jg == N - 1) {
                    u_new[idx(i, j)] = u[idx(i, j)];
                    continue;
                }

                u_new[idx(i, j)] = 0.25 * (
                    u[idx(i - 1, j)] + u[idx(i + 1, j)] +
                    u[idx(i, j - 1)] + u[idx(i, j + 1)]
                );

                double diff = std::abs(u_new[idx(i, j)] - u[idx(i, j)]);
                local_change = std::max(local_change, diff);
            }
        }

        // --- Global Convergence Check ---
        MPI_Allreduce(&local_change, &global_change,
                      1, MPI_DOUBLE, MPI_MAX, comm_cart);

        std::swap(u, u_new);
        iter++;

        if (rank == 0 && iter % 500 == 0) {
            std::cout << "[Iter " << std::setw(5) << iter
                      << "] max_diff = "
                      << std::scientific
                      << global_change << "\n";
        }
    }

    double t1 = MPI_Wtime();

    if (rank == 0) {
        std::cout << "\n"
                  << (global_change <= EPS ? "✓ CONVERGENCE" : "⚠ MAX_ITER")
                  << " after " << iter << " iterations\n"
                  << "  Final change: " << global_change << "\n"
                  << "  Total time: " << std::fixed << std::setprecision(3)
                  << (t1 - t0) << " s\n";
    }

    MPI_Type_free(&col_type);
    MPI_Comm_free(&comm_cart);
    MPI_Finalize();
    return 0;
}

File: PHASE_3/Jacobi_MPI/jacobi_2d_full.f90
! =============================================================
! FULL 2D JACOBI — 2D Decomposition with Cartesian Topology
! =============================================================
! Complete version featuring:
!  - MPI_Cart_create (2D process grid)
!  - MPI_Dims_create (automatic decomposition)
!  - Halo exchange on all 4 sides
!  - Convergence check with MPI_Allreduce
!
! Each process manages a rectangular subdomain.
!
! Compilation:  mpif90 -O2 -Wall -o jacobi2d_f jacobi_2d_full.f90
! Execution:    mpirun -np 4 ./jacobi2d_f
! =============================================================

program jacobi_2d
    use mpi
    implicit none

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! Global parameters
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    integer, parameter :: N       = 128
    real(8), parameter :: EPS     = 1.0d-5
    integer, parameter :: MAX_ITER = 5000

    ! MPI variables
    integer :: rank_world, size_world, ierr
    integer :: rank, coords(2)
    integer :: dims(2), periods(2)
    integer :: comm_cart
    integer :: north, south, west, east, dummy

    ! Subdomain dimensions
    integer :: my_row, my_col
    integer :: local_rows, local_cols
    integer :: global_row_start, global_col_start
    integer :: nr, nc   ! including ghost cells

    ! Grid arrays (1D storage, row-major via manual indexing)
    real(8), allocatable :: u(:,:), u_new(:,:)

    ! Convergence
    real(8)  :: local_change, global_change
    integer  :: iter

    ! MPI datatype for column exchange
    integer :: col_type

    ! Timing
    real(8)  :: t0, t1

    ! Loop indices
    integer :: i, j, ig, jg

    ! Tags
    integer, parameter :: TAG1 = 1, TAG2 = 2, TAG3 = 3, TAG4 = 4

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! MPI Initialization
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    call MPI_Init(ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, rank_world, ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, size_world, ierr)

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! 2D Cartesian topology
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    dims(1)    = 0;  dims(2)    = 0
    periods(1) = 0;  periods(2) = 0

    call MPI_Dims_create(size_world, 2, dims, ierr)
    call MPI_Cart_create(MPI_COMM_WORLD, 2, dims, periods, 1, comm_cart, ierr)
    call MPI_Comm_rank(comm_cart, rank, ierr)
    call MPI_Cart_coords(comm_cart, rank, 2, coords, ierr)

    my_row = coords(1)   ! row position in the process grid
    my_col = coords(2)   ! column position

    local_rows = N / dims(1)
    local_cols = N / dims(2)

    global_row_start = my_row * local_rows
    global_col_start = my_col * local_cols

    if (rank == 0) then
        write(*,'(A,I0,A,I0)') "Global grid: ", N, "x", N
        write(*,'(A,I0,A,I0,A,I0,A)') "Processes: ", dims(1), "x", dims(2), &
                                        " (", size_world, " total)"
        write(*,'(A,I0,A,I0)') "Subdomain per process: ", local_rows, "x", local_cols
        write(*,*)
    end if

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! Neighbors (may be MPI_PROC_NULL on boundaries)
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    call MPI_Cart_shift(comm_cart, 0, -1, dummy, north, ierr)
    call MPI_Cart_shift(comm_cart, 0, +1, dummy, south, ierr)
    call MPI_Cart_shift(comm_cart, 1, -1, dummy, west,  ierr)
    call MPI_Cart_shift(comm_cart, 1, +1, dummy, east,  ierr)

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! Allocation: subdomain + ghost cells
    ! Fortran arrays: u(0:nr+1, 0:nc+1) — 1-based interior [1:nr, 1:nc]
    ! We use explicit ghost rows/cols: u(0:local_rows+1, 0:local_cols+1)
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    nr = local_rows + 2
    nc = local_cols + 2

    allocate( u(0:local_rows+1, 0:local_cols+1) )
    allocate( u_new(0:local_rows+1, 0:local_cols+1) )
    u     = 0.0d0
    u_new = 0.0d0

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! Boundary conditions: bottom boundary = 1.0
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    if (my_row == dims(1) - 1) then
        do j = 1, local_cols
            u(local_rows, j) = 1.0d0
        end do
    end if

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! MPI datatype for columns (not contiguous in Fortran either
    ! with row-major intent; here columns ARE contiguous — stride 1 in dim 1)
    ! In Fortran column-major layout, a column slice u(1:local_rows, j)
    ! is contiguous, so we do NOT need a derived type for columns.
    ! We DO need one for rows (u(i, 1:local_cols) is strided).
    !
    ! North/South exchange  → rows   → strided (MPI_Type_vector)
    ! East/West exchange    → columns → contiguous (plain MPI_DOUBLE)
    !
    ! NOTE: this is the opposite of C++ (row-major).
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

    ! Row type: local_cols elements, stride = local_rows+2 (leading dim)
    call MPI_Type_vector(local_cols, 1, local_rows+2, MPI_DOUBLE_PRECISION, &
                         col_type, ierr)
    call MPI_Type_commit(col_type, ierr)

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! Jacobi loop
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    t0 = MPI_Wtime()
    iter = 0
    global_change = 1.0d0

    do while (global_change > EPS .and. iter < MAX_ITER)

        ! --- North-South Halo Exchange (rows) ---
        ! In Fortran layout, u(i, 1:local_cols) is a row — strided in memory.
        ! We pass u(1,1) / u(local_rows,1) and use col_type to pick local_cols
        ! elements with stride (local_rows+2).

        ! Send row 1 north, receive into ghost row 0
        call MPI_Sendrecv( u(1, 1),          1, col_type, north, TAG1, &
                           u(0, 1),          1, col_type, north, TAG2, &
                           comm_cart, MPI_STATUS_IGNORE, ierr )

        ! Send row local_rows south, receive into ghost row local_rows+1
        call MPI_Sendrecv( u(local_rows,   1), 1, col_type, south, TAG2, &
                           u(local_rows+1, 1), 1, col_type, south, TAG1, &
                           comm_cart, MPI_STATUS_IGNORE, ierr )

        ! --- East-West Halo Exchange (columns, contiguous in Fortran) ---
        ! Send column 1 west, receive into ghost column 0
        call MPI_Sendrecv( u(1, 1),           local_rows, MPI_DOUBLE_PRECISION, west, TAG3, &
                           u(1, local_cols+1), local_rows, MPI_DOUBLE_PRECISION, east, TAG3, &
                           comm_cart, MPI_STATUS_IGNORE, ierr )

        ! Send column local_cols east, receive into ghost column local_cols+1
        call MPI_Sendrecv( u(1, local_cols),  local_rows, MPI_DOUBLE_PRECISION, east, TAG4, &
                           u(1, 0),           local_rows, MPI_DOUBLE_PRECISION, west, TAG4, &
                           comm_cart, MPI_STATUS_IGNORE, ierr )

        ! --- Jacobi Update ---
        local_change = 0.0d0

        do j = 1, local_cols
            jg = global_col_start + (j - 1)
            do i = 1, local_rows
                ig = global_row_start + (i - 1)

                ! Preserve boundary conditions
                if (ig == 0 .or. ig == N-1 .or. jg == 0 .or. jg == N-1) then
                    u_new(i, j) = u(i, j)
                    cycle
                end if

                u_new(i, j) = 0.25d0 * ( &
                    u(i-1, j) + u(i+1, j) + &
                    u(i, j-1) + u(i, j+1) )

                local_change = max(local_change, abs(u_new(i,j) - u(i,j)))
            end do
        end do

        ! --- Global Convergence Check ---
        call MPI_Allreduce(local_change, global_change, 1, &
                           MPI_DOUBLE_PRECISION, MPI_MAX, comm_cart, ierr)

        ! Swap arrays (pointer swap via intrinsic trick or explicit copy)
        u = u_new
        iter = iter + 1

        if (rank == 0 .and. mod(iter, 500) == 0) then
            write(*,'(A,I5,A,ES12.5)') "[Iter ", iter, "] max_diff = ", global_change
        end if

    end do

    t1 = MPI_Wtime()

    if (rank == 0) then
        write(*,*)
        if (global_change <= EPS) then
            write(*,'(A,I0,A)') "✓ CONVERGENCE after ", iter, " iterations"
        else
            write(*,'(A,I0,A)') "⚠ MAX_ITER reached after ", iter, " iterations"
        end if
        write(*,'(A,ES12.5)') "  Final change: ", global_change
        write(*,'(A,F8.3,A)')  "  Total time:   ", (t1 - t0), " s"
    end if

    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    ! Cleanup
    ! ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    deallocate(u, u_new)
    call MPI_Type_free(col_type, ierr)
    call MPI_Comm_free(comm_cart, ierr)
    call MPI_Finalize(ierr)

end program jacobi_2d
